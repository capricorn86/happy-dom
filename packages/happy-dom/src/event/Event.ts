import IEventInit from './IEventInit.js';
import INode from '../nodes/node/INode.js';
import IWindow from '../window/IWindow.js';
import IShadowRoot from '../nodes/shadow-root/IShadowRoot.js';
import IEventTarget from './IEventTarget.js';
import NodeTypeEnum from '../nodes/node/NodeTypeEnum.js';
import { performance } from 'perf_hooks';
import EventPhaseEnum from './EventPhaseEnum.js';
import IDocument from '../nodes/document/IDocument.js';

/**
 * Event.
 */
export default class Event {
	public composed = false;
	public bubbles = false;
	public cancelable = false;
	public defaultPrevented = false;
	public eventPhase: EventPhaseEnum = EventPhaseEnum.none;
	public _immediatePropagationStopped = false;
	public _propagationStopped = false;
	public _target: IEventTarget = null;
	public _currentTarget: IEventTarget = null;
	public timeStamp: number = performance.now();
	public type: string = null;
	public _isInPassiveEventListener = false;
	public NONE = EventPhaseEnum.none;
	public CAPTURING_PHASE = EventPhaseEnum.capturing;
	public AT_TARGET = EventPhaseEnum.atTarget;
	public BUBBLING_PHASE = EventPhaseEnum.bubbling;

	/**
	 * Constructor.
	 *
	 * @param type Event type.
	 * @param [eventInit] Event init.
	 */
	constructor(type: string, eventInit: IEventInit = null) {
		this.type = type;

		if (eventInit) {
			this.bubbles = eventInit.bubbles || false;
			this.cancelable = eventInit.cancelable || false;
			this.composed = eventInit.composed || false;
		}
	}

	/**
	 * Returns target.
	 *
	 * @returns Target.
	 */
	public get target(): IEventTarget {
		return this._target;
	}

	/**
	 * Returns target.
	 *
	 * @returns Target.
	 */
	public get currentTarget(): IEventTarget {
		return this._currentTarget;
	}

	/**
	 * Returns "true" if propagation has been stopped.
	 *
	 * @returns "true" if propagation has been stopped.
	 */
	public get cancelBubble(): boolean {
		return this._propagationStopped;
	}

	/**
	 * Returns composed path.
	 *
	 * @returns Composed path.
	 */
	public composedPath(): IEventTarget[] {
		if (!this._target) {
			return [];
		}

		const composedPath = [];
		let eventTarget: INode | IShadowRoot | IWindow = <INode | IShadowRoot>(<unknown>this._target);

		while (eventTarget) {
			composedPath.push(eventTarget);

			if ((<INode>(<unknown>eventTarget)).parentNode) {
				eventTarget = (<INode>(<unknown>eventTarget)).parentNode;
			} else if (
				this.composed &&
				(<INode>eventTarget).nodeType === NodeTypeEnum.documentFragmentNode &&
				(<IShadowRoot>eventTarget).host
			) {
				eventTarget = (<IShadowRoot>eventTarget).host;
			} else if ((<INode>eventTarget).nodeType === NodeTypeEnum.documentNode) {
				eventTarget = (<IDocument>(<unknown>eventTarget)).defaultView;
			} else {
				break;
			}
		}

		return composedPath;
	}

	/**
	 * Init event.
	 *
	 * @deprecated
	 * @param type Type.
	 * @param [bubbles=false] "true" if it bubbles.
	 * @param [cancelable=false] "true" if it cancelable.
	 */
	public initEvent(type: string, bubbles = false, cancelable = false): void {
		this.type = type;
		this.bubbles = bubbles;
		this.cancelable = cancelable;
	}

	/**
	 * Prevents default.
	 */
	public preventDefault(): void {
		if (!this._isInPassiveEventListener) {
			this.defaultPrevented = true;
		}
	}

	/**
	 * Stops immediate propagation.
	 */
	public stopImmediatePropagation(): void {
		this._immediatePropagationStopped = true;
	}

	/**
	 * Stops propagation.
	 */
	public stopPropagation(): void {
		this._propagationStopped = true;
	}
}
