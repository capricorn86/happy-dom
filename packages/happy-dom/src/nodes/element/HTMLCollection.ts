import IHTMLCollection from './IHTMLCollection.js';

/**
 * HTML collection.
 */
export default class HTMLCollection<T> extends Array implements IHTMLCollection<T> {
	protected _namedItems: { [k: string]: T[] } = {};

	/**
	 * Returns item by index.
	 *
	 * @param index Index.
	 */
	public item(index: number): T | null {
		return index >= 0 && this[index] ? this[index] : null;
	}

	/**
	 * Returns named item.
	 *
	 * @param name Name.
	 * @returns Node.
	 */
	public namedItem(name: string): T | null {
		return this._namedItems[name] && this._namedItems[name].length
			? this._namedItems[name][0]
			: null;
	}

	/**
	 * Appends named item.
	 *
	 * @param node Node.
	 * @param name Name.
	 */
	public _appendNamedItem(node: T, name: string): void {
		if (name) {
			this._namedItems[name] = this._namedItems[name] || [];

			if (!this._namedItems[name].includes(node)) {
				this._namedItems[name].push(node);
			}

			if (!this.hasOwnProperty(name) && this._isValidPropertyName(name)) {
				this[name] = this._namedItems[name][0];
			}
		}
	}

	/**
	 * Appends named item.
	 *
	 * @param node Node.
	 * @param name Name.
	 */
	public _removeNamedItem(node: T, name: string): void {
		if (name && this._namedItems[name]) {
			const index = this._namedItems[name].indexOf(node);

			if (index > -1) {
				this._namedItems[name].splice(index, 1);

				if (this._namedItems[name].length === 0) {
					delete this._namedItems[name];
					if (this.hasOwnProperty(name) && this._isValidPropertyName(name)) {
						delete this[name];
					}
				} else if (this._isValidPropertyName(name)) {
					this[name] = this._namedItems[name][0];
				}
			}
		}
	}

	/**
	 * Returns "true" if the property name is valid.
	 *
	 * @param name Name.
	 * @returns True if the property name is valid.
	 */
	protected _isValidPropertyName(name: string): boolean {
		return (
			!this.constructor.prototype.hasOwnProperty(name) &&
			!Array.prototype.hasOwnProperty(name) &&
			(isNaN(Number(name)) || name.includes('.'))
		);
	}
}
