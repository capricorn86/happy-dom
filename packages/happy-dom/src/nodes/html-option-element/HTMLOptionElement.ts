import IAttr from '../attr/IAttr';
import HTMLElement from '../html-element/HTMLElement';
import IHTMLFormElement from '../html-form-element/IHTMLFormElement';
import HTMLSelectElement from '../html-select-element/HTMLSelectElement';
import INode from '../node/INode';
import IHTMLOptionElement from './IHTMLOptionElement';

/**
 * HTML Option Element.
 *
 * Reference:
 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLOptionElement.
 */
export default class HTMLOptionElement extends HTMLElement implements IHTMLOptionElement {
	public _index: number;
	public _selectedness = false;
	public _dirtyness = false;

	/**
	 * Returns inner text, which is the rendered appearance of text.
	 *
	 * @returns Inner text.
	 */
	public get text(): string {
		return this.innerText;
	}

	/**
	 * Sets the inner text, which is the rendered appearance of text.
	 *
	 * @param innerText Inner text.
	 */
	public set text(text: string) {
		this.innerText = text;
	}

	/**
	 * Returns index.
	 *
	 * @returns Index.
	 */
	public get index(): number {
		return this._index;
	}

	/**
	 * Returns the parent form element.
	 *
	 * @returns Form.
	 */
	public get form(): IHTMLFormElement {
		return <IHTMLFormElement>this._formNode;
	}

	/**
	 * Returns selected.
	 *
	 * @returns Selected.
	 */
	public get selected(): boolean {
		return this._selectedness;
	}

	/**
	 * Sets selected.
	 *
	 * @param selected Selected.
	 */
	public set selected(selected: boolean) {
		const selectNode = <HTMLSelectElement>this._selectNode;

		this._dirtyness = true;
		this._selectedness = Boolean(selected);

		if (selectNode) {
			selectNode._updateOptionItems(this._selectedness ? this : null);
		}
	}

	/**
	 * Returns disabled.
	 *
	 * @returns Disabled.
	 */
	public get disabled(): boolean {
		return this.getAttribute('disabled') !== null;
	}

	/**
	 * Sets disabled.
	 *
	 * @param disabled Disabled.
	 */
	public set disabled(disabled: boolean) {
		if (!disabled) {
			this.removeAttribute('disabled');
		} else {
			this.setAttribute('disabled', '');
		}
	}

	/**
	 * Returns value.
	 *
	 * @returns Value.
	 */
	public get value(): string {
		return this.getAttribute('value') || this.textContent;
	}

	/**
	 * Sets value.
	 *
	 * @param value Value.
	 */
	public set value(value: string) {
		this.setAttribute('value', value);
	}

	/**
	 * @override
	 */
	public setAttributeNode(attribute: IAttr): IAttr {
		const replacedAttribute = super.setAttributeNode(attribute);

		if (
			!this._dirtyness &&
			attribute.name === 'selected' &&
			replacedAttribute?.value !== attribute.value
		) {
			const selectNode = <HTMLSelectElement>this._selectNode;

			this._selectedness = true;

			if (selectNode) {
				selectNode._updateOptionItems(this);
			}
		}

		return replacedAttribute;
	}

	/**
	 * @override
	 */
	public removeAttributeNode(attribute: IAttr): IAttr {
		super.removeAttributeNode(attribute);

		if (!this._dirtyness && attribute.name === 'selected') {
			const selectNode = <HTMLSelectElement>this._selectNode;

			this._selectedness = false;

			if (selectNode) {
				selectNode._updateOptionItems();
			}
		}

		return attribute;
	}

	/**
	 * @override
	 */
	public override _connectToNode(parentNode: INode = null): void {
		const oldSelectNode = <HTMLSelectElement>this._selectNode;

		super._connectToNode(parentNode);

		if (oldSelectNode !== this._selectNode) {
			if (oldSelectNode) {
				oldSelectNode._updateOptionItems();
			}
			if (this._selectNode) {
				(<HTMLSelectElement>this._selectNode)._updateOptionItems();
			}
		}
	}
}
