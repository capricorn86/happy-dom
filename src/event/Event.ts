import Node from '../nodes/basic-types/node/Node';
import IEventInit from './IEventInit';

export default class Event {
	public readonly bubbles: boolean = false;
	public readonly cancelable: boolean = false;
	public readonly composed: boolean = false;
	public readonly currentTarget: Node = null;
	public readonly target: Node = null;
	public defaultPrevented: boolean = false;
	public _immediatePropagationStopped: boolean = false;
	public _propagationStopped: boolean = false;
	public type: string = null;

	/**
	 * Constructor.
	 *
	 * @param {string} type Event type.
	 * @param {IEventInit} eventInit Event init.
	 */
	constructor(type: string, eventInit: IEventInit = null) {
		this.type = type;
		this.bubbles = eventInit && eventInit.bubbles ? true : false;
		this.cancelable = eventInit && eventInit.cancelable ? true : false;
		this.composed = eventInit && eventInit.composed ? true : false;
	}

	/**
	 * Init event.
	 *
	 * @legacy
	 */
	public initEvent(): void {
		this.defaultPrevented = true;
	}

	/**
	 * Prevents default.
	 */
	public preventDefault(): void {
		this.defaultPrevented = true;
	}

	/**
	 * Stops immediate propagation.
	 */
	public stopImmediatePropagation(): void {
		this._immediatePropagationStopped = true;
	}

	/**
	 * Stops propagation.
	 */
	public stopPropagation(): void {
		this._propagationStopped = true;
	}
}
