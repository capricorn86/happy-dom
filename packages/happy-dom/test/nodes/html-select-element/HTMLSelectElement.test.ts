import Window from '../../../src/window/Window.js';
import Document from '../../../src/nodes/document/Document.js';
import HTMLSelectElement from '../../../src/nodes/html-select-element/HTMLSelectElement.js';
import HTMLOptionElement from '../../../src/nodes/html-option-element/HTMLOptionElement.js';
import ValidityState from '../../../src/validity-state/ValidityState.js';
import Event from '../../../src/event/Event.js';
import { beforeEach, describe, it, expect } from 'vitest';
import HTMLCollection from '../../../src/nodes/element/HTMLCollection.js';

describe('HTMLSelectElement', () => {
	let window: Window;
	let document: Document;
	let element: HTMLSelectElement;

	beforeEach(() => {
		window = new Window();
		document = window.document;
		element = <HTMLSelectElement>document.createElement('select');
	});

	describe('Object.prototype.toString', () => {
		it('Returns `[object HTMLSelectElement]`', () => {
			expect(Object.prototype.toString.call(element)).toBe('[object HTMLSelectElement]');
		});
	});

	describe('get options()', () => {
		it('Reflects changes as options elements are added and removed from the DOM.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			option1.value = 'option1';
			element.appendChild(option1);

			expect(element.options.length).toBe(1);
			expect((<HTMLOptionElement>element.options[0]).value).toBe('option1');

			element.removeChild(option1);

			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');
			option2.value = 'option2';
			option3.value = 'option3';
			element.appendChild(option2);
			element.appendChild(option3);

			expect(element.options.length).toBe(2);
			expect((<HTMLOptionElement>element.options[0]).value).toBe('option2');
			expect((<HTMLOptionElement>element.options[1]).value).toBe('option3');
		});
	});

	describe('get value()', () => {
		it('Returns the value of the first option element in the list of options in tree order that has its selectedness set to true.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			option1.selected = true;
			option1.value = 'option1';
			option2.value = 'option2';
			element.appendChild(option1);
			element.appendChild(option2);

			expect(element.value).toBe('option1');
		});

		it('Returns empty string if there are no options.', () => {
			expect(element.value).toBe('');
		});

		it('Returns empty string if no option is selected.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			option1.value = 'option1';
			option2.value = 'option2';
			element.appendChild(option1);
			element.appendChild(option2);

			expect(element.value).toBe('option1');

			element.selectedIndex = -1;

			expect(element.value).toBe('');
		});
	});

	describe('set value()', () => {
		it('Sets options.selectedIndex.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			option1.value = 'option1';
			option2.value = 'option2';
			element.appendChild(option1);
			element.appendChild(option2);

			element.value = 'option1';

			expect(element.options.selectedIndex).toBe(0);
		});

		it('Dispatches "change" event.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			option1.value = 'option1';
			option2.value = 'option2';
			element.appendChild(option1);
			element.appendChild(option2);

			let dispatchedEvent: Event | null = null;
			element.addEventListener('change', (event: Event) => (dispatchedEvent = event));

			element.value = 'option2';

			expect((<Event>(<unknown>dispatchedEvent)).type).toBe('change');

			dispatchedEvent = null;

			element.value = 'option2';

			expect(dispatchedEvent).toBeNull();
		});
	});

	for (const property of ['disabled', 'autofocus', 'required', 'multiple']) {
		describe(`get ${property}()`, () => {
			it('Returns attribute value.', () => {
				expect(element[property]).toBe(false);
				element.setAttribute(property, '');
				expect(element[property]).toBe(true);
			});
		});

		describe(`set ${property}()`, () => {
			it('Sets attribute value.', () => {
				element[property] = true;
				expect(element.getAttribute(property)).toBe('');
			});
		});
	}

	describe(`get name()`, () => {
		it('Returns attribute value.', () => {
			expect(element.name).toBe('');
			element.setAttribute('name', 'value');
			expect(element.name).toBe('value');
		});
	});

	describe(`set name()`, () => {
		it('Sets attribute value.', () => {
			element.name = 'value';
			expect(element.getAttribute('name')).toBe('value');
		});
	});

	describe(`get selectedIndex()`, () => {
		it('Defaults to -1.', () => {
			expect(element.selectedIndex).toBe(-1);
		});

		it('Returns options selectedIndex.', () => {
			element.appendChild(document.createElement('option'));
			element.appendChild(document.createElement('option'));

			element.options.selectedIndex = 1;
			expect(element.selectedIndex).toBe(1);
		});

		it('Returns option with "selected" attribute is defined.', () => {
			const option1 = document.createElement('option');
			const option2 = document.createElement('option');

			option2.setAttribute('selected', '');

			element.appendChild(option1);
			element.appendChild(option2);

			expect(element.selectedIndex).toBe(1);

			option1.setAttribute('selected', '');

			expect(element.selectedIndex).toBe(0);

			option2.removeAttribute('selected');

			expect(element.selectedIndex).toBe(0);
		});
	});

	describe(`get selectedOptions()`, () => {
		it('Defaults to an empty HTMLCollection.', () => {
			expect(element.selectedOptions).toBeInstanceOf(HTMLCollection);
			expect(element.selectedOptions.length).toBe(0);
		});

		it('Returns selected options when "selected" attribute is defined.', () => {
			const option1 = document.createElement('option');
			const option2 = document.createElement('option');

			option2.setAttribute('selected', '');

			element.appendChild(option1);
			element.appendChild(option2);

			expect(element.selectedOptions.length).toBe(1);
			expect(element.selectedOptions[0]).toBe(option2);

			option1.setAttribute('selected', '');

			expect(element.selectedOptions.length).toBe(1);
			expect(element.selectedOptions[0]).toBe(option1);

			option2.removeAttribute('selected');

			expect(element.selectedOptions.length).toBe(1);
			expect(element.selectedOptions[0]).toBe(option1);
		});

		it('Returns selected options when "selected" attribute is defined for multiple options.', () => {
			element.setAttribute('multiple', '');
			const option1 = document.createElement('option');
			const option2 = document.createElement('option');

			option2.setAttribute('selected', '');

			element.appendChild(option1);
			element.appendChild(option2);

			expect(element.selectedOptions.length).toBe(1);
			expect(element.selectedOptions[0] === option2).toBe(true);

			option1.setAttribute('selected', '');

			expect(element.selectedOptions.length).toBe(2);
			expect(element.selectedOptions[0] === option1).toBe(true);
			expect(element.selectedOptions[1] === option2).toBe(true);

			option2.removeAttribute('selected');

			expect(element.selectedOptions.length).toBe(1);
			expect(element.selectedOptions[0] === option1).toBe(true);
		});
	});

	describe('get symbol()', () => {
		it('returns existing symbol properties', () => {
			const symbol = Symbol('test');
			element[symbol] = 'test';
			expect(element[symbol]).toBe('test');
		});

		it('ignores missing symbol properties', () => {
			const symbol = Symbol('other-test');

			expect(element[symbol]).toBe(undefined);

			// https://github.com/capricorn86/happy-dom/issues/1526
			expect(symbol in element).toBe(false);
		});
	});

	describe(`set selectedIndex()`, () => {
		it('Allows -1', () => {
			element.selectedIndex = -1;
		});

		it('Sets options selectedIndex.', () => {
			element.appendChild(document.createElement('option'));
			element.appendChild(document.createElement('option'));

			element.selectedIndex = 1;
			expect(element.options.selectedIndex).toBe(1);
		});

		it('Ignores invalid values gracefully.', () => {
			element.appendChild(document.createElement('option'));
			element.appendChild(document.createElement('option'));

			expect(element.options.selectedIndex).toBe(0);

			element.selectedIndex = <number>(<unknown>undefined);
			expect(element.options.selectedIndex).toBe(0);

			element.selectedIndex = 1000;
			expect(element.options.selectedIndex).toBe(-1);
		});

		it('Dispatched "change" event.', () => {
			const option1 = document.createElement('option');
			const option2 = document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);

			let dispatchedEvent: Event | null = null;
			element.addEventListener('change', (event: Event) => (dispatchedEvent = event));

			element.selectedIndex = 1;

			expect((<Event>(<unknown>dispatchedEvent)).type).toBe('change');

			dispatchedEvent = null;

			element.selectedIndex = 1;

			expect(dispatchedEvent).toBeNull();
		});
	});

	describe(`get labels()`, () => {
		it('Returns associated labels', () => {
			const label1 = document.createElement('label');
			const label2 = document.createElement('label');
			const parentLabel = document.createElement('label');

			label1.setAttribute('for', 'select1');
			label2.setAttribute('for', 'select1');

			element.id = 'select1';

			parentLabel.appendChild(element);
			document.body.appendChild(label1);
			document.body.appendChild(label2);
			document.body.appendChild(parentLabel);

			const labels = element.labels;

			expect(labels.length).toBe(3);
			expect(labels[0] === label1).toBe(true);
			expect(labels[1] === label2).toBe(true);
			expect(labels[2] === parentLabel).toBe(true);
		});
	});

	describe('get validity()', () => {
		it('Returns an instance of ValidityState.', () => {
			expect(element.validity).toBeInstanceOf(ValidityState);
		});
	});

	describe('get validationMessage()', () => {
		it('Returns validation message.', () => {
			element.setCustomValidity('Error message');
			expect(element.validationMessage).toBe('Error message');
		});
	});

	describe(`add()`, () => {
		it('Appends options.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');

			element.add(option1);
			element.add(option2);

			expect(element.length).toBe(2);
			expect(element.children.length).toBe(2);
			expect(element.options.length).toBe(2);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option2).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option2).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option2).toBe(true);
		});

		it('Appends an option before an index.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.add(option1);
			element.add(option2);
			element.add(option3, 1);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option3).toBe(true);
			expect(element[2] === option2).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option3).toBe(true);
			expect(element.children[2] === option2).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option3).toBe(true);
			expect(element.options[2] === option2).toBe(true);
		});

		it('Appends an option before an option element.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.add(option1);
			element.add(option2);
			element.add(option3, option2);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option3).toBe(true);
			expect(element[2] === option2).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option3).toBe(true);
			expect(element.children[2] === option2).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option3).toBe(true);
			expect(element.options[2] === option2).toBe(true);
		});
	});

	describe(`item()`, () => {
		it('Returns an option element on a specified index.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.add(option1);
			element.add(option2);
			element.add(option3);

			expect(element.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option2).toBe(true);
			expect(element.item(2) === option3).toBe(true);
			expect(element.options.item(0) === option1).toBe(true);
			expect(element.options.item(1) === option2).toBe(true);
			expect(element.options.item(2) === option3).toBe(true);
		});
	});

	describe(`appendChild()`, () => {
		it('Adds appended option or option group elements to the HTMLOptionsCollection.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.appendChild(option3);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option2).toBe(true);
			expect(element[2] === option3).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option2).toBe(true);
			expect(element.children[2] === option3).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option2).toBe(true);
			expect(element.options[2] === option3).toBe(true);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option2).toBe(true);
			expect(element.item(2) === option3).toBe(true);
		});

		it('Does not include other types of elements in the HTMLOptionsCollection.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');
			const div = <HTMLOptionElement>document.createElement('div');

			element.appendChild(option1);
			element.appendChild(option2);
			element.appendChild(div);
			element.appendChild(option3);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(4);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option2).toBe(true);
			expect(element[2] === option3).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option2).toBe(true);
			expect(element.children[2] === div).toBe(true);
			expect(element.children[3] === option3).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option2).toBe(true);
			expect(element.options[2] === option3).toBe(true);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option2).toBe(true);
			expect(element.item(2) === option3).toBe(true);
		});

		it('Dispatches "change" event.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			let dispatchedEvent: Event | null = null;

			element.addEventListener('change', (event: Event) => (dispatchedEvent = event));

			element.appendChild(option1);
			element.appendChild(option2);
			element.appendChild(option3);

			expect((<Event>(<unknown>dispatchedEvent)).type).toBe('change');
			expect(element.selectedIndex).toBe(0);

			dispatchedEvent = null;

			option3.selected = true;

			expect((<Event>(<unknown>dispatchedEvent)).type).toBe('change');
			expect(element.selectedIndex).toBe(2);

			dispatchedEvent = null;

			option3.remove();

			expect((<Event>(<unknown>dispatchedEvent)).type).toBe('change');
			expect(element.selectedIndex).toBe(0);
		});

		it('Sets "parentNode" of child elements to the proxy and not the original element.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.appendChild(option3);

			expect(option1.parentNode).toBe(element);
			expect(option2.parentNode).toBe(element);
			expect(option3.parentNode).toBe(element);

			expect(option1.parentElement).toBe(element);
			expect(option2.parentElement).toBe(element);
			expect(option3.parentElement).toBe(element);
		});
	});

	describe(`insertBefore()`, () => {
		it('Adds inserted option or option group elements to the HTMLOptionsCollection at correct index.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.insertBefore(option3, option2);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option3).toBe(true);
			expect(element[2] === option2).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option3).toBe(true);
			expect(element.children[2] === option2).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option3).toBe(true);
			expect(element.options[2] === option2).toBe(true);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option3).toBe(true);
			expect(element.item(2) === option2).toBe(true);
		});

		it('Appends inserted option or option group elements to the HTMLOptionsCollection if referenceNode is null.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.insertBefore(option3, null);

			expect(element.length).toBe(3);
			expect(element.children.length).toBe(3);
			expect(element.options.length).toBe(3);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option2).toBe(true);
			expect(element[2] === option3).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option2).toBe(true);
			expect(element.children[2] === option3).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option2).toBe(true);
			expect(element.options[2] === option3).toBe(true);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option2).toBe(true);
			expect(element.item(2) === option3).toBe(true);
		});

		it('Sets "parentNode" of child elements to the proxy and not the original element.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.insertBefore(option3, option2);

			expect(option1.parentNode).toBe(element);
			expect(option2.parentNode).toBe(element);
			expect(option3.parentNode).toBe(element);

			expect(option1.parentElement).toBe(element);
			expect(option2.parentElement).toBe(element);
			expect(option3.parentElement).toBe(element);
		});
	});

	describe(`removeChild()`, () => {
		it('Removes an option or option group elements from the HTMLOptionsCollection.', () => {
			const option1 = <HTMLOptionElement>document.createElement('option');
			const option2 = <HTMLOptionElement>document.createElement('option');
			const option3 = <HTMLOptionElement>document.createElement('option');

			element.appendChild(option1);
			element.appendChild(option2);
			element.appendChild(option3);

			element.removeChild(option2);

			expect(element.length).toBe(2);
			expect(element.children.length).toBe(2);
			expect(element.options.length).toBe(2);
			expect(element[0] === option1).toBe(true);
			expect(element[1] === option3).toBe(true);
			expect(element.children[0] === option1).toBe(true);
			expect(element.children[1] === option3).toBe(true);
			expect(element.options[0] === option1).toBe(true);
			expect(element.options[1] === option3).toBe(true);
			expect(element.item(0) === option1).toBe(true);
			expect(element.item(1) === option3).toBe(true);
		});
	});

	describe('setCustomValidity()', () => {
		it('Returns validation message.', () => {
			element.setCustomValidity('Error message');
			expect(element.validationMessage).toBe('Error message');
			element.setCustomValidity(<string>(<unknown>null));
			expect(element.validationMessage).toBe('null');
			element.setCustomValidity('');
			expect(element.validationMessage).toBe('');
		});
	});

	for (const method of ['checkValidity', 'reportValidity']) {
		describe(`${method}()`, () => {
			it('Returns "true" if the field is "disabled".', () => {
				const option1 = <HTMLOptionElement>document.createElement('option');
				option1.value = '';
				element.appendChild(option1);

				element.required = true;
				element.disabled = true;

				expect(element[method]()).toBe(true);
			});

			it('Returns "false" if invalid.', () => {
				const option1 = <HTMLOptionElement>document.createElement('option');
				option1.value = '';
				element.appendChild(option1);

				element.required = true;

				expect(element[method]()).toBe(false);
			});

			it('Triggers an "invalid" event when invalid.', () => {
				const option1 = <HTMLOptionElement>document.createElement('option');
				option1.value = '';
				element.appendChild(option1);

				element.required = true;

				let dispatchedEvent: Event | null = null;
				element.addEventListener('invalid', (event: Event) => (dispatchedEvent = event));

				element[method]();

				expect((<Event>(<unknown>dispatchedEvent)).type).toBe('invalid');
			});
		});
	}
});
