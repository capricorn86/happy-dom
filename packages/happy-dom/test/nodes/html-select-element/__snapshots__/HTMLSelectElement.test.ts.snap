// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HTMLSelectElement > constructor() > Matches snapshot. 1`] = `
<select>
  <option>
    Option 1
  </option>
  <option>
    Option 2
  </option>
</select>
`;
